runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -24,192 +21,202 @@
-import android.database.Cursor;
 import android.net.Uri;
-import android.os.AsyncTask;
 import android.os.Bundle;
-import android.os.Environment;
 import android.util.Log;
 
-import java.lang.Thread;
-import java.util.LinkedList;
-
 import junit.framework.Assert;
 
-import org.xwalk.core.XWalkLibraryListener.LibraryStatus;
+import org.xwalk.core.XWalkLibraryInterface.DecompressionListener;
+import org.xwalk.core.XWalkLibraryInterface.DownloadListener;
+import org.xwalk.core.XWalkLibraryInterface.InitializationListener;
 
 /**
- * XWalkActiviy is to support cross package resource loading.
- * It provides method to allow overriding getResources() behavior.
+ * <p>The activity which holds the XWalkView object must extend XWalkActivity.
+ * It’s also important to override the abstract method onXWalkReady() which notify Crosswalk
+ * library is ready to use.</p>
+ *
+ * <p>Unlike embedded mode, the Crosswalk library is not loaded yet at the moment the activity is
+ * created in shared mode. So the user can’t use embedding APIs in onCreate() as usual. All process
+ * using embedding API should be implemented in onXWalkReady().</p>
+ *
+ * <p>onXWalkReady() is always called from XWalkActivity.onStart(), and will be called only once.
+ * It may be triggered following onCreate() if the package of Crosswalk runtime library is
+ * already installed on the device before the launch of the application, or following onRestart()
+ * if the end-user navigates back to this activity after jumped to the download page and
+ * installed the package for the first time.</p>
  */
-public abstract class XWalkActivity extends Activity implements XWalkLibraryListener {
-    private static final String XWALK_APK_NAME = "XWalkRuntimeLib.apk";
-
-    private static final String XWALK_APK_MARKET_URL =
-            "market://details?id=" + XWalkCoreWrapper.XWALK_APK_PACKAGE;
-
+public abstract class XWalkActivity extends Activity {
+    private static final String XWALK_APK_MARKET_URL = "market://details?id=org.xwalk.core";
     private static final String TAG = "XWalkActivity";
 
-    private DownloadTask mDownloadTask;
+    private XWalkLibraryListener mLibraryListener;
     private Dialog mActiveDialog;
-    private XWalkCoreWrapper mCoreWrapper;
-
     private boolean mIsXWalkReady;
-    private boolean mIsVisible;
-    private LinkedList<Object> mReservedObjects;
-    private LinkedList<ReflectMethod> mReservedMethods;
     private String mXWalkApkDownloadUrl;
 
-    @Override
-    protected void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
-
-        mDownloadTask = null;
-        mActiveDialog = null;
-        mCoreWrapper = null;
-
-        mIsXWalkReady = false;
-        mIsVisible = false;
-        mReservedObjects = new LinkedList<Object>();
-        mReservedMethods = new LinkedList<ReflectMethod>();
-
-        XWalkCoreWrapper.reset(null, this);
-    }
+    private static class XWalkLibraryListener
```

Comment:
since XWalkLibraryListener is a private inner-class, we can remove the final modifier now. :-)